ledger/eval_test.go

```diff
@@ -562,6 +562,295 @@ func TestEvalAppAllocStateWithTxnGroup(t *testing.T) {
+	defer l.Close()
+
+	newBlock := bookkeeping.MakeBlock(genesisInitState.Block.BlockHeader)
+	eval, err := l.StartEvaluator(newBlock.BlockHeader, 0)
+	require.NoError(t, err)
+	eval.validate = true
+	eval.generate = false
+	eval.proto = config.Consensus[consensusVersion]
+
+	ops, err := logic.AssembleString(approvalProgram)
+	require.NoError(t, err, ops.Errors)
+	approval := ops.Program
+	ops, err = logic.AssembleString("#pragma version 4\nint 1")
+	require.NoError(t, err)
+	clear := ops.Program
+
+	genHash := genesisInitState.Block.BlockHeader.GenesisHash
+	header := transactions.Header{
+		Sender:      addrs[0],
+		Fee:         minFee,
+		FirstValid:  newBlock.Round(),
+		LastValid:   newBlock.Round(),
+		GenesisHash: genHash,
+	}
+	appcall1 := transactions.Transaction{
+		Type:   protocol.ApplicationCallTx,
+		Header: header,
+		ApplicationCallTxnFields: transactions.ApplicationCallTxnFields{
+			GlobalStateSchema: schema,
+			ApprovalProgram:   approval,
+			ClearStateProgram: clear,
+		},
+	}
+
+	appcall2 := transactions.Transaction{
+		Type:   protocol.ApplicationCallTx,
+		Header: header,
+		ApplicationCallTxnFields: transactions.ApplicationCallTxnFields{
+			ApplicationID: basics.AppIndex(1),
+		},
+	}
+
+	appcall3 := appcall2
+	appcall3.Header.Sender = addrs[1]
+
+	var group transactions.TxGroup
+	group.TxGroupHashes = []crypto.Digest{crypto.HashObj(appcall1), crypto.HashObj(appcall2), crypto.HashObj(appcall3)}
+	appcall1.Group = crypto.HashObj(group)
+	appcall2.Group = crypto.HashObj(group)
+	appcall3.Group = crypto.HashObj(group)
+	stxn1 := appcall1.Sign(keys[0])
+	stxn2 := appcall2.Sign(keys[0])
+	stxn3 := appcall3.Sign(keys[1])
+
+	g := []transactions.SignedTxnWithAD{
+		{
+			SignedTxn: stxn1,
+		},
+		{
+			SignedTxn: stxn2,
+		},
+		{
+			SignedTxn: stxn3,
+		},
+	}
+	txgroup := []transactions.SignedTxn{stxn1, stxn2, stxn3}
+	err = eval.TestTransactionGroup(txgroup)
+	if err != nil {
+		return eval, err
+	}
+	err = eval.transactionGroup(g)
+	return eval, err
+}
+
+func TestProtocolAllowsAppPooling(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	approvalProgram1 := `#pragma version 4
+	txn ApplicationID
+	bz create
+	byte "caller"
+	pop
+	b ok
+create:
+	byte 0x01
+	byte "ZC9KNzlnWTlKZ1pwSkNzQXVzYjNBcG1xTU9YbkRNWUtIQXNKYVk2RzRBdExPakQx"
+	addr DROUIZXGT3WFJR3QYVZWTR5OJJXJCMOLS7G4FUGZDSJM5PNOVOREH6HIZE
+	ed25519verify
+	pop
+ok:
+	int 1`
+
+	approvalProgram2 := `#pragma version 4
+	txn ApplicationID
+	bz create
+	byte "caller"
+	pop
+	b ok
+create:
+	byte "creator"
+	keccak256
+	keccak256
+	keccak256
+	keccak256
+	keccak256
+	keccak256
+	pop
+ok:
+	int 1`
+
+	cases := []struct {
+		prog          string
+		expectedError string
+	}{
+		{approvalProgram1, "dynamic cost budget exceeded, executing ed25519verify: remaining budget is 700 but program cost was 1905"},
+		{approvalProgram2, "dynamic cost budget exceeded, executing keccak256: remaining budget is 700 but program cost was 783"},
+	}
+
+	for _, testCase := range cases {
```

Comment:
I think at one point we wanted to do the protocol checks in `data/transactions/logic/evalStateful_test.go`, but there is some ledger code that also needs to be tested (`prepareEvalParams`) for correct protocol, so I put this test in the ledger tests.

Reply:
these tests look good but I think we should not test everything here because in ledger package we only set pooled budget for multiple app calls in one group. And the decisions are made in logic.
Let's do the following:
1.  a unit test for`prepareEvalParams` call on a group with 2 txns with app calls on old and new protos. No ledger construction/blocks needed.
2. Preserve a high-level test with block evaluator: no pooling on old proto and pooling on new proto. Programs should be simple but valid, something like `"int 1" + strings.Repeat("keccak256", 50)`.
3. Make more precise testing on `ed25519verify` in logic package - it is needed anyway, and the `EvalParams` with proto and budget can be easily set there.

Reply:
I added the `prepareEvalParams` tests above like you have suggested. Just to be clear, we should probably only check for correct pooling here, and check for cost/budget in the `data/transactions/logic/evalStateful_test.go` file?

Reply:
It is always a good idea to check both happy case and error case. The idea is to have 1) minimal sufficient coverage 2) without knowing to much details about stuff from other packeges.
I I suggest to use the following code to generate test programs of difference cost by varying N:
```golang
prog := "byte "a\n" + strings.Repeat("keccak256\n", N) + "pop\nint 1\n";
```
So here in `ledger` the following four cases would be enough:
1. cost ~701 + cost ~100 => FAIL on v29.
2. same combo OK on vFuture
3. ~701 + ~701 => FAIL on both v29 and vFuture